staging/src/k8s.io/kube-apiextensions-server/test/integration/registration_test.go

```diff
@@ -338,3 +346,153 @@ func TestDeRegistrationAndReRegistration(t *testing.T) {
 		}
 	}()
 }
+
+func TestEtcdStorage(t *testing.T) {
+	config, err := testserver.DefaultServerConfig()
+	if err != nil {
+		t.Fatal(err)
+	}
+	stopCh, apiExtensionClient, clientPool, err := testserver.StartServer(config)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer close(stopCh)
+
+	etcdPrefix := getPrefixFromConfig(t, config)
+
+	ns1 := "another-default-is-possible"
+	curletDefinition := testserver.NewCurletCustomResourceDefinition(apiextensionsv1alpha1.ClusterScoped)
+	curletVersionClient, err := testserver.CreateNewCustomResourceDefinition(curletDefinition, apiExtensionClient, clientPool)
+	if err != nil {
+		t.Fatal(err)
+	}
+	curletNamespacedResourceClient := NewNamespacedCustomResourceClient(ns1, curletVersionClient, curletDefinition)
+	if _, err := instantiateCustomResource(t, testserver.NewCurletInstance(ns1, "bar"), curletNamespacedResourceClient, curletDefinition); err != nil {
+		t.Fatalf("unable to create curlet cluster scoped Instance:%v", err)
+	}
+
+	ns2 := "the-cruel-default"
+	noxuDefinition := testserver.NewNoxuCustomResourceDefinition(apiextensionsv1alpha1.NamespaceScoped)
+	noxuVersionClient, err := testserver.CreateNewCustomResourceDefinition(noxuDefinition, apiExtensionClient, clientPool)
+	if err != nil {
+		t.Fatal(err)
+	}
+	noxuNamespacedResourceClient := NewNamespacedCustomResourceClient(ns2, noxuVersionClient, noxuDefinition)
+	if _, err := instantiateCustomResource(t, testserver.NewNoxuInstance(ns2, "foo"), noxuNamespacedResourceClient, noxuDefinition); err != nil {
+		t.Fatalf("unable to create noxu namespace scoped Instance:%v", err)
+	}
+
+	testcases := map[string]struct {
+		etcdPath       string
+		expectedObject *metaObject
+	}{
+		"namespacedNoxuDefinition": {
+			etcdPath: path.Join("/", etcdPrefix, "apiextensions.k8s.io/customresourcedefinitions/noxus.mygroup.example.com"), // TODO: Double check this, no namespace?
+			expectedObject: &metaObject{
+				Kind:       "CustomResourceDefinition",
+				APIVersion: "apiextensions.k8s.io/v1alpha1",
+				Metadata: Metadata{
+					Name:      "noxus.mygroup.example.com",
+					Namespace: "",
+					SelfLink:  "/apis/apiextensions.k8s.io/v1alpha1/customresourcedefinitions/noxus.mygroup.example.com",
+				},
+			},
+		},
+		"namespacedNoxuInstance": {
+			etcdPath: path.Join("/", etcdPrefix, "mygroup.example.com/noxus/the-cruel-default/foo"),
+			expectedObject: &metaObject{
+				Kind:       "WishIHadChosenNoxu",
+				APIVersion: "mygroup.example.com/v1alpha1",
+				Metadata: Metadata{
+					Name:      "foo",
+					Namespace: "the-cruel-default",
+					SelfLink:  "", // TODO double check: empty?
```

Comment:
filled in by the REST layer.  In theory /apis can change (used to be the case anyway)